File: rust/pypasswords/src/tests/test_cbc.rs
// Copyright (c) 2025-2026 Arista Networks, Inc.
// Use of this source code is governed by the Apache License 2.0
// that can be found in the LICENSE file.

use super::*;
use crate::errors::CbcDecryptPyError;
use crate::errors::CbcEncryptPyError;

#[test]
fn cbc_decrypt_invalid_base64_err() {
    setup();
    pyo3::Python::attach(|py| {
        let module = py.import("passwords").unwrap();
        let args = ("passwd", "ThisIsNotBase64!!!");

        let err = module.call_method1("cbc_decrypt", args).unwrap_err();

        assert!(err.is_instance_of::<passwords::CBCInvalidBase64Error>(py));
        assert!(err.is_instance_of::<passwords::PasswordError>(py));
        assert_eq!(err.value(py).to_string(), "Invalid Base64 encoding.");
    });
}

#[test]
fn cbc_invalid_base64_error_uses_public_module_path() {
    with_passwords_module(|py, _module| {
        let module_name: String = py
            .get_type::<passwords::CBCInvalidBase64Error>()
            .getattr("__module__")
            .unwrap()
            .extract()
            .unwrap();

        assert_eq!(module_name, "pyavd_utils.passwords");
    });
}

#[test]
fn cbc_decrypt_failed_err() {
    with_passwords_module(|py, module| {
        // "YWJjZA==" is "abcd" (4 bytes).
        // Triple DES requires multiples of 8.
        // This will trigger the .map_err(|_| CbcError::DecryptionFailed) branch.
        let args = ("any_key", "YWJjZA==");

        let err = module.call_method1("cbc_decrypt", args).unwrap_err();

        assert!(err.is_instance_of::<passwords::CBCDecryptionFailedError>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Decryption failed (check password)."
        );
    });
}

#[test]
fn cbc_decrypt_invalid_signature_err() {
    with_passwords_module(|py, module| {
        // Provide valid base64, but no Arista signature at the beginning
        let args = ("some_key", "YWFhYWFhYWFhYWFhYWFhYQ==");
        let err = module.call_method1("cbc_decrypt", args).unwrap_err();

        assert!(err.is_instance_of::<passwords::CBCInvalidSignatureError>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Invalid Arista signature in decrypted data."
        );
    });
}

#[test]
fn cbc_verify_returns_bool() {
    with_passwords_module(|_py, module| {
        let key = "42.42.42.42";
        let data = "arista";

        let encrypted: String = module
            .call_method1("cbc_encrypt", (key, data))
            .unwrap()
            .extract()
            .unwrap();

        // Success case
        let is_valid: bool = module
            .call_method1("cbc_verify", (key, encrypted.clone()))
            .unwrap()
            .extract()
            .unwrap();
        assert!(is_valid);

        // Failure case
        let is_invalid: bool = module
            .call_method1("cbc_verify", ("wrong_key", encrypted))
            .unwrap()
            .extract()
            .unwrap();
        assert!(!is_invalid);
    });
}

#[test]
fn cbc_internal_errors_map_to_specific_pyerrs() {
    with_passwords_module(|py, _module| {
        let err = pyo3::PyErr::from(CbcDecryptPyError::from(::passwords::CbcError::InvalidUtf8));
        assert!(err.is_instance_of::<passwords::CBCInvalidUtf8Error>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Decrypted data is not valid UTF-8."
        );

        let encryption_err = pyo3::PyErr::from(CbcEncryptPyError::from(
            ::passwords::CbcError::EncryptionFailed,
        ));
        assert!(encryption_err.is_instance_of::<passwords::CBCEncryptionFailedError>(py));
        assert_eq!(
            encryption_err.value(py).to_string(),
            "Encryption failed: internal block alignment error."
        );
    });
}

File: tests/passwords/test_cbc.py
# Copyright (c) 2025-2026 Arista Networks, Inc.
# Use of this source code is governed by the Apache License 2.0
# that can be found in the LICENSE file.

from contextlib import AbstractContextManager
from contextlib import nullcontext as does_not_raise

import pytest

from pyavd_utils.passwords import (
    CBCDecryptionFailedError,
    CBCError,
    CBCInvalidBase64Error,
    CBCInvalidSignatureError,
    CBCInvalidUtf8Error,
    PyAVDUtilsPasswordError,
    cbc_decrypt,
    cbc_encrypt,
    cbc_verify,
)


def test_cbc_error_hierarchy() -> None:
    """Test that CBC errors inherit from the passwords base error."""
    assert issubclass(CBCError, PyAVDUtilsPasswordError)
    assert issubclass(CBCInvalidBase64Error, CBCError)


CBC_ENCRYPT_TEST_DATA = [
    pytest.param(
        "42.42.42.42_passwd",
        "arista",
        "3QGcqpU2YTwKh2jVQ4Vj/A==",
        id="Valid encryption",
    ),
]


@pytest.mark.parametrize(("key", "data", "expected_b64"), CBC_ENCRYPT_TEST_DATA)
def test_cbc_encrypt_success(key: str, data: str, expected_b64: str) -> None:
    """Test cbc_encrypt."""
    assert cbc_encrypt(key, data) == expected_b64


CBC_DECRYPT_TEST_DATA = [
    pytest.param(
        "42.42.42.42_passwd",
        "3QGcqpU2YTwKh2jVQ4Vj/A==",
        "arista",
        does_not_raise(),
        id="Valid decryption",
    ),
    pytest.param(
        "any_key",
        "NotBase64!!!",
        "",
        pytest.raises(CBCInvalidBase64Error, match="Invalid Base64 encoding"),
        id="Invalid base64 input",
    ),
    pytest.param(
        "wrong_password",
        "bM7t58t04qSqLHAfZR/Szg==",
        "",
        pytest.raises(CBCInvalidSignatureError, match="Invalid Arista signature"),
        id="Wrong password (signature mismatch)",
    ),
    pytest.param(
        "any_key",
        "YWJjZA==",
        "",
        pytest.raises(CBCDecryptionFailedError, match="Decryption failed"),
        id="Block size / Alignment failure",
    ),
    pytest.param(
        "42.42.42.42_passwd",
        "Sh5yjV8SD2j//////////9pkhd5VI3SbQDy17ujMdko=",
        "",
        pytest.raises(CBCInvalidUtf8Error, match="Decrypted data is not valid UTF-8"),
        id="Invalid UTF-8 sequence in decrypted data",
    ),
]


@pytest.mark.parametrize(("key", "encrypted_data", "expected_plain", "expected_raise"), CBC_DECRYPT_TEST_DATA)
def test_cbc_decrypt(key: str, encrypted_data: str, expected_plain: str, expected_raise: AbstractContextManager[None]) -> None:
    """Test cbc_decrypt susccess and various failures."""
    with expected_raise:
        assert cbc_decrypt(key, encrypted_data) == expected_plain


CBC_VERIFY_TEST_DATA = [
    pytest.param(
        "42.42.42.42_passwd",
        "3QGcqpU2YTwKh2jVQ4Vj/A==",
        True,
        id="Verify success",
    ),
    pytest.param(
        "wrong_password",
        "LIi7vE5hcmlzdGEAAAA=",
        False,
        id="Verify failure (wrong password)",
    ),
    pytest.param(
        "any_key",
        "NotBase64!!!",
        False,
        id="Verify failure (invalid base64)",
    ),
]


@pytest.mark.parametrize(("key", "encrypted_data", "expected_bool"), CBC_VERIFY_TEST_DATA)
def test_cbc_verify(key: str, encrypted_data: str, expected_bool: bool) -> None:
    """Test cbc_verify."""
    assert cbc_verify(key, encrypted_data) == expected_bool
